### ISSUE TITLE
Optimise performance of descendant segment

### ISSUE THREAD

**Issue by glyn** [glyn/jsonpath-nondeterminism#2]
[One of the examples](https://github.com/glyn/jsonpath-nondeterminism/blob/9f3d5d2a5d478b9535abd1ffdc9d17067bc5d755/app/Main.hs#L892) and [a corresponding test](https://github.com/glyn/jsonpath-nondeterminism/blob/9f3d5d2a5d478b9535abd1ffdc9d17067bc5d755/test/Spec.hs#L44), when uncommented, each take several minutes to run because they involve the descendant segment. This is probably because the current algorithm computes all (12!, i.e. over 400 million) permutations of descendants and then filters this collection according to the rules in the RFC.

This inefficient approach was taken because it fairly clearly conforms to the RFC.

Is there an effective optimisation which also fairly clearly conforms to the RFC?

To reproduce uncomment the example and issue:
```
stack run
```
or uncomment the test and issue:
```
stack test
```
Note: the expected value in the test is incorrect and the test should fail.

**Comment by glyn** [comment]
Changing:
~~~
l :: Nodelist <- map (map snd) $ filter validDescendantOrdering $ permutations $ ([], n):descendants [] n
~~~
to:
~~~
l :: Nodelist <- map (map snd) $ map (([], n):) $ filter validDescendantOrdering $ permutations $ descendants [] n
~~~
improved the speed of the problematic example by a factor of 8.

**Comment by glyn** [comment]
Merged and tidied up: 94eac8304bf4d2a3bef130c5086a53ebc8979249
